providers/gpgpu/bin/gpu_passthrough.py

```diff
@@ -0,0 +1,576 @@
+def build_gpu_test(instance: LXD, vendor: str):
+    """Builds the GPU passthrough test."""
+    test_name = ""
+    cmake_cmd = "cmake ../mixbench-{test_name}"
+
+    # Add necessary APT repositories to instance
+    for repo in GPU_VENDORS[vendor].get("repos", []):
+        add_apt_repo(instance, repo)
+
+    if vendor == "nvidia":
+        logging.debug("Installing CUDA Toolkit on instance")
+        packages = ["build-essential", "cmake", "cuda-toolkit"]
+        instance.run(
+            "apt-get -q install -y --no-install-recommends {}".format(
+                " ".join(packages)
+            ),
+            on_guest=True,
+            check=True,
+        )
+
+        logging.debug("Finding CUDA capability for GPU")
+        cuda_arch = "native"
+        proc = instance.run(
+            "nvidia-smi --query-gpu=compute_cap --format=csv,noheader",
+            on_guest=True,
+        )
+        if proc.returncode == 0 and proc.stdout:
+            cuda_arch = proc.stdout.strip().replace(".", "")
+        logging.debug("Using CUDA architecture '%s'", cuda_arch)
+
+        test_name = "cuda"
+        nvcc_path = "/usr/local/cuda/bin/nvcc"
+        cmake_cmd = "CUDACXX={} {} -DCMAKE_CUDA_ARCHITECTURES={}".format(
+            nvcc_path, cmake_cmd, cuda_arch
+        )
+
+    logging.info("Fetching and compiling GPU passthrough test on instance")
+    cmds = [
+        "set -e",
+        "git clone https://github.com/ekondis/mixbench.git",
+        "mkdir mixbench/build-{}".format(test_name),
+        "cd mixbench/build-{}".format(test_name),
+        cmake_cmd.format(test_name=test_name),
+        "make",
+        "ln -s ~/mixbench/build-{0}/mixbench-{0} ~/test".format(test_name),
+    ]
+    instance.run(
+        "bash -c '{}'".format("; ".join(cmds)), on_guest=True, check=True
+    )
+
+
+def run_gpu_test(
+    instance: LXD,
+    run_count: int = GPU_RUNS,
+    threshold_sec: float = GPU_THRESHOLD_SEC,
+):
+    """Executes GPU passthrough test."""
+    logging.info("Running GPU passthrough test %d times", run_count)
+    total_runtime_sec = 0.0
+    for i in range(run_count):
+        tic = time.time()
+        # XXX: If we package the test, this line needs to be updated
+        instance.run("./test", on_guest=True, check=True)
+        toc = time.time()
+        runtime_sec = toc - tic
+        total_runtime_sec += runtime_sec
+        logging.debug("Runtime #%d (sec): %f", i, runtime_sec)
+
+    avg_runtime_sec = total_runtime_sec / run_count
+    logging.info("Average runtime (sec): %f", avg_runtime_sec)
+    if avg_runtime_sec >= threshold_sec:
+        logging.error(
+            "Average runtime %fs greater than threshold %fs",
+            avg_runtime_sec,
+            threshold_sec,
+        )
+        raise SystemExit(1)
+
+
+def test_lxd_gpu(args):
+    """Tests GPU passthrough with a LXD container."""
+    logging.info("Executing LXD GPU passthrough test")
+
+    instance = LXD(args.template, args.rootfs)
+    try:
```

Comment:
This makes me wonder if both LXD and LXDVM should be context managers, that would make it easier to read this and use the functionality in general. Something like
```
with LXD(...) as instance: 
    ...
```

Reply:
Oh, that would be nice. I'll `try` to do that (pun intended)